Update tini install snippet

Firefishy opened the pull request:
* Update tini to v0.18.0 (current latest release)
* Add gpg recv-keys retry loop per https://github.com/docker-library/official-images/issues/4252
* Switch to tini-static instead of dynamic build for automatic compatibility with alpine
* Use `tini --version` instead of `tini --help` which is not available in TINI_MINIMAL releases.

Snippet tested on:
* Ubuntu 12.04 to 18.04
* Debian
* Alpine

tianon commented on README.md:
```diff
@@ -188,16 +188,23 @@ Here is a snippet of a `Dockerfile` to add in tini (be sure to use it in `CMD` o
+ENV TINI_VERSION v0.18.0
+RUN set -ex \
+	&& curl -fSL --compressed "https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini-static" -o /usr/local/bin/tini \
+	&& curl -fSL --compressed "https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini-static.asc" -o /usr/local/bin/tini.asc \
```
In Alpine, `tini` has been available as a package since Alpine 3.4, so I don't think it's very prudent for us to show how to install a static version just for Alpine's sake (since `apk add --no-cache tini` is going to be superior there in almost every way except the ability to get the latest bleeding edge, which is totally sane as an exercise for the reader).

tianon replied:
Why does this include `--compressed` now?  IMO we should instead just switch to `wget` which is easier to get right out-of-the-box (no `-f` needed, for example):

```bash
wget -O /usr/local/bin/tini "https://...."; \
```

tianon commented on README.md:
```diff
@@ -188,16 +188,23 @@ Here is a snippet of a `Dockerfile` to add in tini (be sure to use it in `CMD` o
+				gpg --batch --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys "$key" || \
+				gpg --batch --keyserver hkp://ipv4.pool.sks-keyservers.net --recv-keys "$key" || \
+				gpg --batch --keyserver hkp://pgp.mit.edu:80 --recv-keys "$key" || \
+				gpg --batch --keyserver ha.pool.sks-keyservers.net --recv-keys "$key" ; \
```
This keyserver looping (and looping over a single key) is unnecessary for the official images build infrastructure and definitely not something we'd recommend (see https://github.com/docker-library/official-images/issues/4252#issuecomment-405749749, https://github.com/docker-library/php/pull/666, https://github.com/tianon/pgp-happy-eyeballs).

tianon commented on README.md:
```diff
@@ -188,16 +188,23 @@ Here is a snippet of a `Dockerfile` to add in tini (be sure to use it in `CMD` o
+			done \
 	&& gpg --batch --verify /usr/local/bin/tini.asc /usr/local/bin/tini \
-	&& rm -r "$GNUPGHOME" /usr/local/bin/tini.asc \
+	&& rm -rf "${GNUPGHOME}" /usr/local/bin/tini.asc \
```
I'm not sure the `{}` add any value here?  Also, we should probably include `gpgconf --kill all` before this line (or `command -v gpgconf && gpgconf --kill all || :`, if we want to be hyper-explicit).

tianon commented on README.md:
```diff
@@ -188,16 +188,23 @@ Here is a snippet of a `Dockerfile` to add in tini (be sure to use it in `CMD` o
-	&& curl -fSL "https://github.com/krallin/tini/releases/download/$TINI_VERSION/tini" -o /usr/local/bin/tini \
-	&& curl -fSL "https://github.com/krallin/tini/releases/download/$TINI_VERSION/tini.asc" -o /usr/local/bin/tini.asc \
+ENV TINI_VERSION v0.18.0
+RUN set -ex \
```
We should probably go all the way if we're updating this and do `set -eux;` and use semicolons instead since their `set -e` behavior is much more reliable (since `&&` creates a "complex expression" and the `set -e` rules become less predictable, especially with things like `for` loops embedded).

Firefishy commented:
Thank you for the awesome detailed PR comments. I have updated based on the comments.

* Added notes in snipped to use distro tini if available
* Switched to `wget`
* Removed loop and used standard `||` logic (Bonus: works with centos `/bin/sh` now too)
* Used `gpgconf --kill all` logic and removed the `-f` on `rm`
* `set -eux` now and switched to using `;`

Updated snipped tested on:
* Ubuntu 12.04 to 18.10
* Debian
* Alpine
* Centos 6+

Firefishy commented:
@tianon would you mind reviewing when you have a moment.

yosifkit commented on README.md:
```diff
@@ -184,20 +184,31 @@ Some examples at the time of writing:
-Here is a snippet of a `Dockerfile` to add in tini (be sure to use it in `CMD` or `ENTRYPOINT` as appropriate):
+Be sure to use tini in `CMD` or `ENTRYPOINT` as appropriate.
+
+It is best to install tini from a package. If tini is not available as a package here is a snippet of a `Dockerfile` to add in tini:
```
```suggestion
It is best to install tini from a distribution-provided package (ex. `apt-get install tini`). If tini is not available in your distribution or is too old, here is a snippet of a `Dockerfile` to add in tini:
```

yosifkit commented on README.md:
```diff
@@ -184,20 +184,31 @@ Some examples at the time of writing:
-	&& rm -r "$GNUPGHOME" /usr/local/bin/tini.asc \
-	&& chmod +x /usr/local/bin/tini \
-	&& tini -h
+# alpine:3.4+ rather use: apk add --no-cache tini
```
Rather than maintaining a list of which distributions have `tini` available, we should just stick to the "get it as a package".
```suggestion
```

yosifkit commented on README.md:
```diff
@@ -184,20 +184,31 @@ Some examples at the time of writing:
-	&& chmod +x /usr/local/bin/tini \
-	&& tini -h
+# alpine:3.4+ rather use: apk add --no-cache tini
+# ubuntu:18.10+ rather use: apt-get install -y tini
```
```suggestion
```

yosifkit commented on README.md:
```diff
@@ -184,20 +184,31 @@ Some examples at the time of writing:
-	&& tini -h
+# alpine:3.4+ rather use: apk add --no-cache tini
+# ubuntu:18.10+ rather use: apt-get install -y tini
+# debian:10+ rather use: apt-get install -y tini
```
```suggestion
```

yosifkit commented on README.md:
```diff
@@ -184,20 +184,31 @@ Some examples at the time of writing:
+# alpine:3.4+ rather use: apk add --no-cache tini
+# ubuntu:18.10+ rather use: apt-get install -y tini
+# debian:10+ rather use: apt-get install -y tini
+#
```
```suggestion
```

yosifkit commented on README.md:
```diff
@@ -184,20 +184,31 @@ Some examples at the time of writing:
+  wget -O /usr/local/bin/tini "https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini"; \
+  wget -O /usr/local/bin/tini.asc "https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini.asc"; \
+  export GNUPGHOME="$(mktemp -d)"; \
+  gpg --batch --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys "$TINI_SIGN_KEY" || \
```
This is just, the loop written out; for official-images we only require one server (`ha.pool.sks-keyservers.net`), since we run a proxy to ensure it fails less often. It is basically https://github.com/docker-library/php/pull/666.
```suggestion
```

yosifkit commented on README.md:
```diff
@@ -184,20 +184,31 @@ Some examples at the time of writing:
+  wget -O /usr/local/bin/tini.asc "https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini.asc"; \
+  export GNUPGHOME="$(mktemp -d)"; \
+  gpg --batch --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys "$TINI_SIGN_KEY" || \
+  gpg --batch --keyserver hkp://ipv4.pool.sks-keyservers.net --recv-keys "$TINI_SIGN_KEY" || \
```
```suggestion
```

yosifkit commented on README.md:
```diff
@@ -184,20 +184,31 @@ Some examples at the time of writing:
+  export GNUPGHOME="$(mktemp -d)"; \
+  gpg --batch --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys "$TINI_SIGN_KEY" || \
+  gpg --batch --keyserver hkp://ipv4.pool.sks-keyservers.net --recv-keys "$TINI_SIGN_KEY" || \
+  gpg --batch --keyserver hkp://pgp.mit.edu:80 --recv-keys "$TINI_SIGN_KEY" || \
```
```suggestion
```

Firefishy commented:
@yosifkit Thank you for the review. I have merged and squashed all the recommended changes.
